// @progfay/scrapbox-parser/lib/block/node/DecorationNode.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DecorationNodeParser = void 0;
const _1 = require(".");
const creator_1 = require("./creator");
const PlainNode_1 = require("./PlainNode");
const decorationRegExp = /\[[!"#%&'()*+,\-./{|}<>_~]+ (?:\[[^[\]]+\]|[^\]])+\]/;
const createDecorationNode = (raw, opts) => {
    if (opts.context === "table") {
        return (0, PlainNode_1.createPlainNode)(raw, opts);
    }
    const separatorIndex = raw.indexOf(" ");
    const rawDecos = raw.substring(1, separatorIndex);
    const text = raw.substring(separatorIndex + 1, raw.length - 1);
    const decoSet = new Set(rawDecos);
    if (decoSet.has("*")) {
        const asteriskCount = rawDecos.split("*").length - 1;
        decoSet.delete("*");
        decoSet.add(`*-${Math.min(asteriskCount, 10)}`);
    }
    return [
        {
            type: "decoration",
            raw,
            rawDecos,
            decos: Array.from(decoSet),
            nodes: (0, _1.convertToNodes)(text, { ...opts, nested: true }),
        },
    ];
};
exports.DecorationNodeParser = (0, creator_1.createNodeParser)(createDecorationNode, {
    parseOnNested: false,
    parseOnQuoted: true,
    patterns: [decorationRegExp],
});
